orderbook/src/bad_token.rs

```diff
@@ -0,0 +1,465 @@
+    async fn find_address_owning_token(&self, token: H160) -> Result<Option<(H160, U256)>> {
+        const BATCH_SIZE: usize = 100;
+
+        let pairs = self
+            .base_tokens
+            .iter()
+            .filter_map(move |&base_token| TokenPair::new(base_token, token));
+        let candidates = pairs
+            .flat_map(|pair| self.pools.iter().map(move |pool| pool.pair_address(&pair)))
+            .collect::<HashSet<_>>();
+
+        let instance = ERC20::at(&self.web3, token);
+        let mut batch = CallBatch::new(self.web3.transport());
+        let futures = candidates
+            .iter()
+            .map(|&address| {
+                let fut = instance.balance_of(address).batch_call(&mut batch);
+                async move { (address, fut.await) }
+            })
+            .collect::<Vec<_>>();
+        batch.execute_all(BATCH_SIZE).await;
+
+        let mut biggest_balance = None;
+        for future in futures {
+            let (address, result) = future.await;
+            let balance = match result {
+                Ok(balance) => balance,
+                Err(err) => {
+                    return match EthcontractError::classify(&err) {
+                        EthcontractError::Node => Err(err.into()),
+                        EthcontractError::Contract => Ok(None),
+                    }
+                }
+            };
+            match biggest_balance {
+                Some((_, current_biggest)) if current_biggest > balance => (),
+                _ => biggest_balance = Some((address, balance)),
+            }
+        }
+        Ok(biggest_balance)
+    }
+
+    fn create_trace_request(&self, token: H160, amount: U256, take_from: H160) -> Vec<CallRequest> {
+        let instance = ERC20::at(&self.web3, token);
+
+        // We transfer one unit (base on decimals) of the token from the amm pool into the
+        // settlement contract and then to an arbitrary address.
+        // Note that gas use can depend on the recipient because for the standard implementation
+        // sending to an address that does not have any balance yet (implicitly 0) causes an
+        // allocation.
+
+        let mut requests = Vec::new();
+
+        // 0
+        let tx = instance.balance_of(self.settlement_contract).m.tx;
+        requests.push(call_request(None, token, tx));
+        // 1
+        let tx = instance.transfer(self.settlement_contract, amount).tx;
+        requests.push(call_request(Some(take_from), token, tx));
+        // 2
+        let tx = instance.balance_of(self.settlement_contract).m.tx;
+        requests.push(call_request(None, token, tx));
+        // 3
+        // For the out transfer we use an arbitrary address without balance to detect tokens that
+        // usually apply fees but not if the the sender or receiver is specifically exempt like
+        // their own uniswap pools.
+        let recipient = H160::from_low_u64_le(1);
```

Comment:
We should make this:
```rust
H160(hex!("cd7328a5D376D5530f054EAF0B9D235a4Fd36059"));
```

Which is:
```rust
PrivateKey::from_hex(keccak256("moo")).unwrap().public_address();
```

(This is mostly a joke and for fun :stuck_out_tongue:).

Reply:
good idea